src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -136,22 +160,55 @@ private Task FlushPendingWritesAsync(TKey key, CancellationToken cancellationTok
             private Task AddWriteTaskAsync(TKey key, Action<SqlConnection> action, CancellationToken cancellationToken)
                 => Storage.AddWriteTaskAsync(_writeQueueKeyToWrites, GetWriteQueueKey(key), action, cancellationToken);
 
-            private Stream ReadBlob(SqlConnection connection, TDatabaseId dataId)
+            private Stream ReadBlob(
+                SqlConnection connection, TDatabaseId dataId, string columnName,
+                Checksum checksumOpt, CancellationToken cancellationToken)
             {
+                // Note: it's possible that someone may write to this row between when we
+                // get the row ID above and now.  That's fine.  We'll just read the new
+                // bytes that have been written to this location.  Note that only the
+                // data for a row in our system can change, the ID will always stay the
+                // same, and the data will always be valid for our ID.  So there is no
+                // safety issue here.
                 if (TryGetRowId(connection, dataId, out var rowId))
```

Comment:
any reason we don't run this under transaction so that we can stay safe when we finally support delete on db?

Reply:
there is no concept (or need) to delete row-ids.

Reply:
still not sure why we don't get that under transaction? also, is RowId in a table never re-used?

Reply:
so it is 2 step to read any blob data. first need to read row id from same table from data id, and then read blob from it. and we need to read everything from it at once.

Reply:
still not sure what kind of different putting TryGetRowId inside of transaction and not make?

Reply:
> so it is 2 step to read any blob data. first need to read row id from same table from data id, and then read blob from it. and we need to read everything from it at once.

Correct.  One is the index, one is the data.  

> still not sure what kind of different putting TryGetRowId inside of transaction and not make?

It is not necessary to be in a transaction, so it isn't in the transaction.  I don't want the code to be written to do things that are unnecessary.  It could lead people to make incorrect assumptions about things.  The index doesn't need to be in the transaction section.  It is designed to be 100% safe regardless of how people read/write from it.  i.e. you should not be able to observe corrupt data, nor can you race against anyone with it.  That is baked heavily into the impl to guarantee that.

This is contrasted with blob data.  Prior to this change, blob data could be seen to be corrupt if you didn't have transactions, so transactions were only used for that purpose.  After this cahnge, in order to prevent races with checksums reads of checksum+data must all be in transactions.

I only want to have the code in transactions that are *absolutely necessary* for correctness.

Reply:
how people know? what is safe to be done outside of transaction and what is not? so, only you can modify this code? what bad thing happen if transaction is covering all thing? will it leak memory like what happened before for.. conenction I think?

Reply:
> how people know? what is safe to be done outside of transaction and what is not? 

How do you know what is safe to be done outside of a lock, and what is not?  Ideally with documentation and knowledge transfer.  I've documented in the code *why* we need transactions.  For example: 

```c#
            // NOTE: we do need to do the blob reading in a transaction because of the
            // following: https://www.sqlite.org/c3ref/blob_open.html
            //
            // If the row that a BLOB handle points to is modified by an UPDATE, DELETE, 
            // or by ON CONFLICT side-effects then the BLOB handle is marked as "expired".
            // This is true if any column of the row is changed, even a column other than
            // the one the BLOB handle is open on. Calls to sqlite3_blob_read() and 
            // sqlite3_blob_write() for an expired BLOB handle fail with a return code of
            // SQLITE_ABORT.
```

and

```
                    // Have to run the blob reading in a transaction.  This is necessary
                    // for two reasons.  First, blob reading outside a transaction is not
                    // safe to do with the sqlite API.  It may produce corrupt bits if 
                    // another thread is writing to the blob.  Second, if a checksum was
                    // passed in, we need to validate that the checksums match.  This is
                    // only safe if we are in a transaction and no-one else can race with
                    // us.
```

I've attempted to document pretty clearly what the concerns are here.  If you feel more documentation is appropriate, i'm happy to add it.